src/plugins/mvcc_delete_plugin.cpp

```diff
@@ -0,0 +1,177 @@
+}
+
+/**
+ * This function analyzes each chunk of every table and triggers a chunk-cleanup-procedure if a certain threshold of invalidated rows is exceeded.
+ */
+void MvccDeletePlugin::_logical_delete_loop() {
+  for (auto& [table_name, table] : _sm.tables()) {
+    // Check all chunks, except for the last one, which is currently used for insertions
+    if (table->chunk_count() == 1) return;
+    ChunkID max_chunk_id_to_check = static_cast<ChunkID>(table->chunk_count() - 2);
+
+    for (ChunkID chunk_id = ChunkID{0}; chunk_id <= max_chunk_id_to_check; chunk_id++) {
+      const auto& chunk = table->get_chunk(chunk_id);
+      if (chunk && chunk->get_cleanup_commit_id() == MvccData::MAX_COMMIT_ID) {
+        // Calculate metric 1 – Chunk invalidation level
+        const double rate_of_invalidated_rows = static_cast<double>(chunk->invalid_row_count()) / chunk->size();
+
+        // Calculate metric 2 – Chunk Hotness
+        const CommitID lowest_end_commit_id =
+            *std::min_element(std::begin(chunk->mvcc_data()->end_cids), std::end(chunk->mvcc_data()->end_cids));
+        const CommitID commit_id_diff = TransactionManager::get().last_commit_id() - lowest_end_commit_id;
+        const CommitID max_commit_id_diff =
+            table->max_chunk_size() * static_cast<CommitID>(_delete_threshold_commit_diff_factor);
+
+        // Evaluate metrics
+        const bool criterion1 = _delete_threshold_rate_invalidated_rows <= rate_of_invalidated_rows;
+        const bool criterion2 = max_commit_id_diff <= commit_id_diff;
+        if (criterion1 || criterion2) _delete_chunk(table_name, chunk_id);
+      }
+    }  // for each chunk
+  }    // for each table
+}
+
+/**
+ * This function processes the physical-delete-queue until its empty.
+ */
+void MvccDeletePlugin::_physical_delete_loop() {
+  std::unique_lock<std::mutex> lock(_mutex_physical_delete_queue);
+
+  while (!_physical_delete_queue.empty()) {
+    ChunkSpecifier chunk_spec = _physical_delete_queue.front();
+    bool success = _delete_chunk_physically(chunk_spec.table_name, chunk_spec.chunk_id);
+
+    if (success) {
+      _physical_delete_queue.pop();
+    } else {
+      return;  // wait for more transactions to finish
+    }
+  }
+}
+
+void MvccDeletePlugin::_delete_chunk(const std::string& table_name, const ChunkID chunk_id) {
+  // Delete chunk logically
+  bool success = _delete_chunk_logically(table_name, chunk_id);
+
+  // Queue physical delete
+  if (success) {
+    DebugAssert(StorageManager::get().get_table(table_name)->get_chunk(chunk_id)->get_cleanup_commit_id() !=
+                    MvccData::MAX_COMMIT_ID,
+                "Chunk needs to be deleted logically before deleting it physically.")
+
+        std::unique_lock<std::mutex>
+            lock(_mutex_physical_delete_queue);
+    _physical_delete_queue.emplace(table_name, chunk_id);
+  } else {
+    std::cout << "Logical delete of chunk " << chunk_id << " failed." << std::endl;
```

Comment:
I think it would be better if this throws an exception.

Reply:
We actually don't want this to throw an exception, because we are expecting transactions to fail, since it's a natural part of the optimistic transaction management approach. That's why we would just rollback our Tx and try it again.

Reply:
Maybe write "failed because of MVCC conflict, retrying"?

Reply:
Done